### PR TITLE
nix-lib: convert to a proper license in callTemplate

### PR DESCRIPTION
**Pull request by sternenseemann**
Take naive lookup code from `yarn2nix-moretea`: Lookup license SPDX identifiers in `lib.licenses` and convert appropriately.

~~It doesn't support SPDX Expression syntax, but that'd be a project for a separate nix library, implementing stuff like `(A OR (B WITH EXCEPTION))`.~~ Implementing SPDX Expression syntax doesn't seem to be possible with how licenses work in nixpkgs currently.

### CODE REVIEW

**Comment by sternenseemann** [comment]
We also could include more data from `package.json` while allowing the user to override:

* `name`
* `version`

**Comment by sternenseemann** [comment]
Are utility functions at the bottom of the file? Then I'll move `spdxLicense` down to where `formatKey` is.

**Comment by Profpatsch** [comment]
SGTM

**Comment by Profpatsch** [comment]
Problem is that nix-community/yarn2nix is GPL3, so we can’t just copy this snippet verbatim I think.

**Comment by sternenseemann** [comment]
Moved the function. 

License shouldn't be a problem since the file is (afaik) not even part of the nix-community repository. I copied it out of `nixpkgs` which is MIT, so it should be at least dual-licensed.

File: nix-lib/default.nix [hunk 1]
```diff
@@ -115,6 +116,22 @@ let
...
     else "@${scope}/${name}";
 
+  # Helper function to convert from a SPDX license string as found
+  # in package.json (https://docs.npmjs.com/files/package.json#license)
+  # to a nix license attrSet as found in lib.licenses.
+  #
+  # Implementation is taken from yarn2nix-moretea.
+  # It does not support SPDX expression syntax.
+  # https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix#L35-L42
+  spdxLicense = licstr:
```
Comment (Profpatsch) [inline nix-lib/default.nix]: I’d call it something more longform, like `npmLicenseToNixpkgsSpdxLicense`

Comment (sternenseemann) [reply nix-lib/default.nix]: :+1:

**Comment by Profpatsch** [comment]
Now that we have a CI/test setup, do you think we can add a test?

**Comment by Profpatsch** [comment]
I don’t think this will work, since the `package.json` license field can contain basically arbitrary things: https://docs.npmjs.com/files/package.json#license

We only parse string license fields at the moment. The license could also be a combination of multiple SPDX licenses …

There is a Haskell parser for the SPDX DSL in the Cabal library: 
- https://hackage.haskell.org/package/Cabal-3.2.0.0/docs/Distribution-SPDX.html
- https://hackage.haskell.org/package/spdx 

Maybe we should do some advanced magic to figure out the nix licenses. However, it’s probably best to just use the SPDX string verbatim …

**Comment by sternenseemann** [comment]
On the other hand, a lot of node packages will have a plain SPDX string as their `license` field which we can easily lookup in `lib.licenses`. `UNLICENSED` is something we should definitely map to `licenses.unfree`.

Mapping all other license expressions to `{ shortName = npmLicense; }` is probably a bad idea and in this case we should just pass them along as a string.

Another headache is that the npm `license` field *isn't even required to be a spdx license expression*. It can be arbitrary strings like `"You can find our license at <https://rms.sexy>"`.

**Comment by sternenseemann** [comment]
Finally added some tests. Also changed the semantics of `npmSpdxToNixpkgsLicense` a bit: Now, if we don't find a match, we just leave the value as-is instead of stuffing it in a set with a `shortName` attribute.

**Comment by sternenseemann** [comment]
With the direction we are currently going with IfD, `callTemplate` won't be a central chokepoint in the library we can use for data processing. We should probably implement the license conversion in Haskell, even though it means we need to eval nix expressions, even part of `nixpkgs` using HNIX.

**Pull request closed without merging**